### PR TITLE
rewrite BindingMethodOverridesPass because it was wrong

### PR DESCRIPTION
**Pull request by bbrehm**
It used to mark methods as overridden even if they were not actually overriden. Also, the test coverage was empty.

### CODE REVIEW

File: semanticcpg/src/main/scala/io/shiftleft/semanticcpg/passes/BindingMethodOverridesPass.scala [hunk 1]
```diff
@@ -1,48 +1,53 @@
...
-    diff.build()
+      }
+    }
+    for (typeDecl <- cpg.typeDecl.toIterator;
+         binding <- typeDecl.bindsOut.asScala) {
+      diffGraph.addNodeProperty(node = binding,
+                                key = NodeKeyNames.IS_METHOD_NEVER_OVERRIDDEN,
+                                value = (!overwritten.contains(binding)).asInstanceOf[AnyRef])
+    }
+    return Some(diffGraph.build()).iterator
```
Comment (ml86) [inline semanticcpg/src/main/scala/io/shiftleft/semanticcpg/passes/BindingMethodOverridesPass.scala:41]: Just `Iterator(diffGraph.build())`

File: semanticcpg/src/main/scala/io/shiftleft/semanticcpg/passes/BindingMethodOverridesPass.scala [hunk 2]
```diff
@@ -1,48 +1,53 @@
...
   }
 
-  def isMethodNeverOverridden(typeDecl: nodes.TypeDecl, bindingName: String, bindingSignature: String): Boolean = {
-    typeDecl.start.derivedTypeDeclTransitive.methodBinding
-      .filter(_.nameExact(bindingName).signatureExact(bindingSignature))
-      .headOption
-      .isEmpty
+  def markRecurse(binding: nodes.Binding): Unit = {
+    if (overwritten.add(binding)) {
```
Comment (ml86) [inline semanticcpg/src/main/scala/io/shiftleft/semanticcpg/passes/BindingMethodOverridesPass.scala:45]: Please give the return value of `add` a proper name for better readability.

File: semanticcpg/src/main/scala/io/shiftleft/semanticcpg/passes/BindingMethodOverridesPass.scala [hunk 3]
```diff
@@ -1,48 +1,54 @@
...
-import io.shiftleft.passes.{CpgPass, DiffGraph, ParallelIteratorExecutor}
+import io.shiftleft.codepropertygraph.generated.{nodes, NodeKeyNames}
+import io.shiftleft.passes.{CpgPass, DiffGraph}
 import io.shiftleft.semanticcpg.language._
-import org.apache.logging.log4j.LogManager
+import scala.jdk.CollectionConverters._
 
 class BindingMethodOverridesPass(cpg: Cpg) extends CpgPass(cpg) {
-  import BindingMethodOverridesPass._
+  val overwritten = scala.collection.mutable.HashSet[nodes.Binding]()
```
Comment (Ferada) [inline semanticcpg/src/main/scala/io/shiftleft/semanticcpg/passes/BindingMethodOverridesPass.scala:10]:
```suggestion
  val overwritten = mutable.HashSet[nodes.Binding]()
```

AFAIK that's how we have the imports elsewhere and it's not as long as the full version.

File: semanticcpg/src/main/scala/io/shiftleft/semanticcpg/passes/BindingMethodOverridesPass.scala [hunk 4]
```diff
@@ -1,48 +1,54 @@
...
+    for (typeDecl <- cpg.typeDecl.toIterator) {
+      val parentTypeDecls = typeDecl.inheritsFromOut.asScala.map {
+        _.refOut.next
+      }.toList
+      for (binding <- typeDecl.bindsOut.asScala) {
+        if (!overwritten.contains(binding)) {
+          val method = binding.refOut.next
+          for (parentTypeDecl <- parentTypeDecls) {
+            val parentBinding = bindingTable.get((binding.name, binding.signature, parentTypeDecl))
+            if (parentBinding.isDefined && parentBinding.get.refOut.next != method) {
```
Comment (Ferada) [inline semanticcpg/src/main/scala/io/shiftleft/semanticcpg/passes/BindingMethodOverridesPass.scala:28]:
```suggestion
            parentBinding.foreach { ... }
```
At least that way don't have to spell out both `isDefined` and `get`.

File: semanticcpg/src/main/scala/io/shiftleft/semanticcpg/passes/BindingMethodOverridesPass.scala [hunk 5]
```diff
@@ -1,48 +1,54 @@
...
+    for (typeDecl <- cpg.typeDecl.toIterator;
+         binding <- typeDecl.bindsOut.asScala) {
+      bindingTable.update((binding.name, binding.signature, typeDecl), binding)
+    }
+    for (typeDecl <- cpg.typeDecl.toIterator) {
+      val parentTypeDecls = typeDecl.inheritsFromOut.asScala.map {
+        _.refOut.next
+      }.toList
+      for (binding <- typeDecl.bindsOut.asScala) {
+        if (!overwritten.contains(binding)) {
```
Comment (Ferada) [inline semanticcpg/src/main/scala/io/shiftleft/semanticcpg/passes/BindingMethodOverridesPass.scala:24]: I'd have maybe said `if (overwritten.contains(binding)) continue` just to reduce the nesting level a bit.

File: semanticcpg/src/main/scala/io/shiftleft/semanticcpg/passes/BindingMethodOverridesPass.scala [hunk 6]
```diff
@@ -1,48 +1,54 @@
...
+              markRecurse(parentBinding.get)
+            }
+          }
         }
-      })
-    diff.build()
+      }
+    }
+    for (typeDecl <- cpg.typeDecl.toIterator;
+         binding <- typeDecl.bindsOut.asScala) {
```
Comment (Ferada) [inline semanticcpg/src/main/scala/io/shiftleft/semanticcpg/passes/BindingMethodOverridesPass.scala:36]: Since this loop happened up there already, I guess they can't / shouldn't be merged or cached?

File: semanticcpg/src/main/scala/io/shiftleft/semanticcpg/passes/BindingMethodOverridesPass.scala (hunk 5 above)
Comment (bbrehm) [reply semanticcpg/src/main/scala/io/shiftleft/semanticcpg/passes/BindingMethodOverridesPass.scala:24]: Yes!

I'd love to do this! Unfortunately scala has neither continue nor break, because it is run by fanatics that put misconceived notions of purity before pragmatic needs of users. 

I'll spare you the extended rant about this language. I have given up "elegant scala", and have settled for "would have been barely readable, if not for scalafmt messing up the formatting".

File: semanticcpg/src/main/scala/io/shiftleft/semanticcpg/passes/BindingMethodOverridesPass.scala (hunk 4 above)
Comment (bbrehm) [reply semanticcpg/src/main/scala/io/shiftleft/semanticcpg/passes/BindingMethodOverridesPass.scala:28]: Afaiu `foreach` opens a stackframe / closure. That's more complex and therefore harder to read than a simple conditional branch.

File: semanticcpg/src/main/scala/io/shiftleft/semanticcpg/passes/BindingMethodOverridesPass.scala (hunk 6 above)
Comment (bbrehm) [reply semanticcpg/src/main/scala/io/shiftleft/semanticcpg/passes/BindingMethodOverridesPass.scala:36]: Yeah, you're right: Theoretically, the "Officially Correct Way (TM)" would be to iterate in topologically sorted order, with respect to inheritance relationship; then we could merge some of the loops. But I think this would be too big of a gun / complicated, and there shouldn't be that many bindings / typeDecls.

I assumed that looping ` for (typeDecl <- cpg.typeDecl.toIterator;
         binding <- typeDecl.bindsOut.asScala)` was roughly as fast as caching the looped-over items.
